fendermint/vm/topdown/src/finality.rs

```diff
@@ -0,0 +1,459 @@
+        let v = self.parent_view_data.validator_set(height)?;
+        Ok(v)
+    }
+
+    fn top_down_msgs(&self, height: BlockHeight) -> StmDynResult<Vec<CrossMsg>> {
+        let v = self.parent_view_data.top_down_msgs(height, height)?;
+        Ok(v)
+    }
+
+    fn new_parent_view(
+        &self,
+        height: BlockHeight,
+        block_hash: BlockHash,
+        validator_set: ValidatorSet,
+        top_down_msgs: Vec<CrossMsg>,
+    ) -> StmDynResult<()> {
+        if !top_down_msgs.is_empty() {
+            // make sure incoming top down messages are ordered by nonce sequentially
+            ensure_sequential_by_nonce(&top_down_msgs)?;
+        };
+
+        let r = self.parent_view_data.height_data.modify(|mut cache| {
+            let r = cache
+                .append(height, (block_hash, validator_set, top_down_msgs))
+                .map_err(Error::NonSequentialParentViewInsert);
+            (cache, r)
+        })?;
+
+        if let Err(e) = r {
+            return abort(e);
+        }
+
+        Ok(())
+    }
+}
+
+impl ParentFinalityProvider for InMemoryFinalityProvider {
+    fn last_committed_finality(&self) -> StmDynResult<IPCParentFinality> {
+        let finality = self.last_committed_finality.read_clone()?;
+        Ok(finality)
+    }
+
+    fn next_proposal(&self) -> StmDynResult<Option<IPCParentFinality>> {
+        let latest_height = if let Some(h) = self.parent_view_data.latest_height()? {
+            h
+        } else {
+            return abort(Error::HeightNotReady);
+        };
+
+        // latest height has not reached, we should wait or abort
+        if latest_height < self.config.chain_head_delay {
+            return abort(Error::HeightThresholdNotReached);
+        }
+
+        let height = latest_height - self.config.chain_head_delay;
+        let last_committed_finality = self.last_committed_finality.read()?;
+
+        // parent height is not ready to be proposed yet
+        if height <= last_committed_finality.height {
+            return Ok(None);
+        }
+
+        // prepare block hash and validator set
+        let height_data = self.parent_view_data.height_data.read()?;
+        let block_hash = if let Some(v) = height_data.get_value(height) {
+            let (block_hash, _, _) = v;
+            block_hash.clone()
+        } else {
+            return abort(Error::HeightNotFoundInCache(height));
+        };
+
+        Ok(Some(IPCParentFinality { height, block_hash }))
+    }
+
+    fn check_proposal(&self, proposal: &IPCParentFinality) -> StmDynResult<()> {
+        self.check_height(proposal)?;
+        self.check_block_hash(proposal)
+    }
+
+    fn on_finality_committed(&self, finality: &IPCParentFinality) -> StmDynResult<()> {
+        // the height to clear
+        let height = finality.height;
+
+        self.parent_view_data.height_data.update(|mut cache| {
+            cache.remove_key_below(height + 1);
```

Comment:
I think this is wrong, we shouldn't garbage collect when the finality is committed, but when the membership and message information is executed, as we may need to query that information to perform the execution. May be we can do it at the same time as `on_finality_committed` but I would either decouple both processes so is clear, all document this function (especially the garbage collection side of things) well.

Reply:
When the proposal is executed, it will execute the messages first, and once they are executed, we will call `on_finality_committed` so that previous cache is cleared. I assume `membership and message information is executed` and `the finality is committed` will happen in one function call.